packages/drivers/odsp-driver/src/odspDocumentService.ts

```diff
@@ -288,157 +308,126 @@ export class OdspDocumentService implements IDocumentService {
                     tenantId,
                     documentId,
                     token,
                     io,
                     client,
-                    url2!,
+                    afdUrl,
                     20000,
                     this.logger,
-                ).then((connection) => {
-                    // Refresh AFD cache
-                    const cacheResult = this.writeLocalStorage(lastAfdConnectionTimeMsKey, Date.now().toString());
-                    if (cacheResult) {
-                        // eslint-disable-next-line max-len
-                        debug(`Cached AFD connection time. Expiring in ${new Date(Number(localStorage.getItem(lastAfdConnectionTimeMsKey)) + afdUrlConnectExpirationMs)}`);
-                    }
-                    logger.sendPerformanceEvent({
-                        eventName: "UsedAfdUrl",
-                        duration: endNonAfd - startNonAfd,
-                        refreshedCache: cacheResult,
-                        fromCache: false,
-                    }, connectionError);
-
-                    return connection;
-                }).catch((retryError) => {
-                    logger.sendPerformanceEvent({
-                        eventName: "FailedAfdUrlFallback",
-                        duration: endNonAfd - startNonAfd,
-                    }, retryError);
-                    throw retryError;
+                );
+                // Set the successful connection attempt in the cache so we can skip the non-AFD failure the next time
+                // we try to connect and immediately try AFD instead.
+                writeLocalStorage(lastAfdConnectionTimeMsKey, Date.now().toString());
+                this.logger.sendTelemetryEvent({
+                    eventName: "AfdConnectionSuccess",
                 });
-            } else {
-                logger.sendPerformanceEvent({
-                    eventName: "FailedNonAfdUrl-NoAfdFallback",
-                }, connectionError);
+                return connection;
+            } catch (connectionError) {
+                // Clear cache since it failed
+                localStorage.removeItem(lastAfdConnectionTimeMsKey);
+                // Log before throwing
+                const canRetry = canRetryOnError(connectionError);
+                this.logger.sendTelemetryEvent(
+                    {
+                        eventName: "AfdConnectionFail",
+                        canRetry,
+                    },
+                    connectionError,
+                );
+                throw connectionError;
+            }
+        };
+
+        const afdCacheValid = isAfdCacheValid();
+
+        // First use the AFD URL if we've logged a successful AFD connection in the cache - its presence in the cache
+        // means the non-AFD url has failed in the past, in which case we would prefer to skip doing another
+        // attempt->fail on the non-AFD.
+        if (afdCacheValid && afdUrl !== undefined) {
```

Comment:
I'd personally change it. I'd rather create a "plan" of connections in the form of uris: [string], and then just execute it.
I'm not a fan of so many telemetry events & verboseness - I'd rather see error object (or error properties object) being populated as we populate the plan, execute the plan, and then being thrown at the end.

Reply:
Latest change reduces emissions to one per connection attempt, emitted when the attempt succeeds or fails.  Since we don't have a way to emit multiple events with a single telemetry submission, I think this is minimal verbosity while still achieving the goal of this PR of rounding out the missing telemetry.